### DISCUSSION TITLE
Raids: How should they work with Owncast?

### DISCUSSION THREAD

**Discussion by gabek** [owncast/owncast#717] [Ideas]
For those unfamiliar, on Twitch a ["Raid"](https://help.twitch.tv/s/article/how-to-use-raids?language=en_US) is when a streamer ends their broadcast and sends all of their viewers to another stream.

To have something with a similar feel, but in a decentralized manner, it'll take some thought.

For clarity let's call the original stream Stream A, and the stream you want to Raid to Stream B.

## Considerations 

1. We don't want to just embed Stream B's video + chat into Stream A's video + chat UI, as I think it's important for people to view the actual server of a streamer and give the streamer a "real click", and the viewer the full experience of that streamer's server.  That's why we don't just embed the video of people's streams in the directory, for example.
2. We don't want to (and likely can't due to browser restrictions) programatically navigate the user to Stream B's server URL.
3. It should take into account the privacy of Stream B and the fact that they may not want a Raid.
4. We only support raiding to servers that use HTTPS.

Here's a flow I came up with that I think takes all of this into account, but is by no means the perfect solution.  Maybe it's a good jumping off point, though.  As a note, some of this would use the similar functionality on the frontend that I was thinking about for #679, a feature to embed 3rd party content into the Owncast UI.

1. Stream A, via the Admin, says they want to Raid Stream B, and puts in the URL to Stream B's server.
2. Stream A's server makes an API request to Stream B's endpoint of `/api/raid`, and that endpoint simply returns back if it supports raids or not.  It does not take any action, and it's not a secured endpoint in any way.
3. If Stream B returns back to Stream A that it does not support Raids then show a message in Stream A's Admin that Raids are not supported for Stream B.
4. If Stream B supports Raids then a system chat message on Stream A is sent telling everyone that they're raiding Stream B.
5. A modal is presented with an iframe (similar to the idea of #679) that includes Stream B's server.  This modal should be easy to close so users can exit it and labeled to obviously state what's going on.

While it would be nice to have a way to tell Stream B "You're being Raided" so it can put something in Stream B's chat, or perform some kind of action, there's no secure way that's straightforward where people couldn't just spam that endpoint and fake a ton of raid events.  But if we really wanted this, we could figure something out that works similarly to the YP protocol where there's a multi-step/multi-request verification, but I kind of prefer the servers being completely decoupled and not having to talk to each other to pull this off.

What do you think?

**Comment by gingervitis** [comment]
Here's a rough mock of what it could look like.

1. Server A (Owncast Demo server)  initializes a "raid" to Server B (w0rmh0le live! server) from their Admin.  
2. Assuming that Server B's url is valid, and is currently streaming, when Server A's stream ends, it will first display a message in the chat.
<img width="1092" alt="image" src="https://user-images.githubusercontent.com/9563255/107844277-e6ca6080-6d86-11eb-943a-27c851087c27.png">

3. After a few seconds, Server A's page could display a large size `<iframe>` with Server B's stream in it, along with a message on top indicating that Server B's content has been loaded into the frame. 

4. The message notice can be dismissed with the X button, at which point, the iframe could then either take up the 100% height and width of the browser (while still being at Server A's url), or just actually redirect you to Server B's url to get out of iframe clunkiness.




<img width="1400" alt="image" src="https://user-images.githubusercontent.com/9563255/107843885-b33a0700-6d83-11eb-83d6-52ec46907361.png">

(mobile)
<img width="321" alt="image" src="https://user-images.githubusercontent.com/9563255/107844091-69522080-6d85-11eb-9a0b-8a09f42954dd.png">

**Reply by geekgonecrazy** [reply]
Ok that looks pretty awesome visually! 😍

**Comment by geekgonecrazy** [comment]
Could we issue an intent to raid from Stream A to Stream B that does nothing other then emit an event for Stream B's integrations?

I'm not sure any more at risk for spam then sending random messages in the chat?

If we wanted to get a bit more complicated.. we could let people add a list of approved raiders if they want to prevent spam.

**Reply by gabek** [reply]
I guess the difference is we have plans to lock down the websocket to try to stop spam messages.  Where in this case it'd be a plan to open up an endpoint to anybody who wants to send fake raid events to any owncast server they want.

**Comment by gabek** [comment]
Just came up with an idea!

Like discussed, we create an endpoint that any server can send messages to, to say that Stream A is raiding Stream B, _however_, Stream B, in the admin, will get some kind of alert or "pending raids" queue, and they have to acknowledge it.  If they don't, then it'll just sit there unacknowledged, and no message will get announced saying that Stream A raided Stream B.

It doesn't stop Stream A from raiding B if B doesn't acknowledge it (though it could work that way), but it would stop any kind of actions from firing.

The downside is somehow we'd have to make it super obvious that the admin has a pending raid that Server B has to acknowledge.  But if server B were to get spammed, it can't be so annoying that the streamer on B would get flooded with popups or something.

**Reply by controlfreakstudio** [reply]
I like an announced hand-off/raid much but do you envisage this also doing the x is hosting y ala twitch or just a hand-off. and need moar amigos on the platform before this is too pressing

**Reply by geekgonecrazy** [reply]
This could only be owncast to owncast I think.  Interop to twitch wouldn't really work. 

But I think the idea here is to discuss how it could work for the future when more people are here and this does become more useful :)

**Comment by aaronpk** [comment]
What's the reason for using an iframe instead of doing an actual HTTP redirect to the raided server? I don't really see the benefit in UX or functionality of popping up the iframe. The iframe method will also break if the next server raids another server.

Also, as soon as the chat identity stuff is sorted out, showing the player in an iframe will break all the cookies needed for that to work, so it'd be a much more future-proof solution to do an actual redirect instead.

**Comment by flaki** [comment]
You'll definitely need some sort of 'identity handoff' for raiding to work across servers.

RE:
> The message notice can be dismissed with the X button, at which point, the iframe could then either take up the 100% height and width of the browser (while still being at Server A's url), or just actually redirect you to Server B's url to get out of iframe clunkiness.

Funnily enough there is an actual web standard in the works called "[Portals](https://github.com/WICG/portals/blob/main/README.md)", which aims to solve exactly this UX issue! (you can [watch a quick explainer here](https://youtu.be/Z6wjUOSh9Tk?t=405)). I'm not saying we should wait until that lands in all browsers (it's currenly experimental/behind a flag in Chrome), but that provides a potential future upgrade path for an iframe-based implementation.

**Reply by geekgonecrazy** [reply]
Wow!  Mind = blown.  That is a sweet feature.

**Comment by jprjr** [comment]
For the server-to-server portion, you could probably cut a lot of spam by enforcing a few rules:

* ServerA (the raider) needs to have a PTR record on it's IP address
* Said PTR should point to a record that resolves back to the original IP

Another idea is to perform a dial back mechanism.

So, every server would have a public and private key, these could be permanent or generated on-demand for a raid.

ServerA reaches out to serverB with server A's public key, a piece of data, and a callback endpoint.

ServerB holds the request open (maybe?), takes that piece of data, appends it's own data, encrypts it with serverA's public key, and hits the callback URL with said data and serverB's public key

ServerA decrypts, and re-encrypts serverB's data hunk with said public key, and signs it.

ServerB confirms it can decrypt it, and verifies the signature, and let's everything proceed
